drivers: kscan: ft5336 hogs resources in interrupt mode

dhoove opened the pull request:
Allow the interrupt line to be used to correctly detect quick
press/release events, while using the timer to sample at a
configurable rate between the press/down and release/up events,
reducing resource use.

Fixes #33994

Signed-off-by: David Hoover <39233766+dhoove@users.noreply.github.com>

VenkatKotakonda commented on drivers/kscan/Kconfig.ft5336:
```diff
@@ -14,14 +14,23 @@ if KSCAN_FT5336
+config KSCAN_FT5336_INTERRUPT_POLLING_MODE
+	bool "Enable interrupt polling mode"
+	depends on KSCAN_FT5336_INTERRUPT
+	default 10
```
polling_mode is just a flag. this shouldn't be bool?
anyway period is defined above KSCAN_FT5336_PERIOD

VenkatKotakonda replied:
yes, you need to remove this.
usually, KConfig changes should have separate commit and c code files should have separate commit.

dhoove commented:
I'm using custom hardware, but I just created this board file for the STM32L476 Nucleo board:

[board.overlay.nucleol476rg.txt](https://github.com/zephyrproject-rtos/zephyr/files/6280990/board.overlay.nucleol476rg.txt)

#define KSCAN_DEV_NAME DT_LABEL( DT_ALIAS( kscan0 ) )
#include <drivers/kscan.h>

static void kscan_callback( const struct device* dev, uint32_t x, uint32_t y, bool pressed )
{
    LOG_INF( "%d, %d, %d.", x, y, pressed );
}

void main(void)
{
    const struct device* kscan_dev = device_get_binding( KSCAN_DEV_NAME );
    kscan_config( kscan_dev, kscan_callback );
    kscan_enable_callback( kscan_dev );
}

dhoove commented on drivers/kscan/kscan_ft5336.c:
```diff
@@ -72,8 +93,9 @@ struct ft5336_data {
-	/** Timer (polling mode). */
+#endif
+#ifdef KSCAN_FT5336_ENABLE_TIMER
+	/** Timer (polling mode or interrupt polling mode). */
```
Hi. I'm not sure I understand. I haven't changed this Doxygen comment, which refers to the following timer instantiation, except by adding some text inside the parentheses.

VenkatKotakonda replied:
Agree, this one of doxygen comment format. I missed it

jfischer-no commented on drivers/kscan/kscan_ft5336.c:
```diff
@@ -35,6 +41,21 @@ LOG_MODULE_REGISTER(ft5336, CONFIG_KSCAN_LOG_LEVEL);
 /* REG_Pn_XH: Position */
 #define POSITION_H_MSK		0x0FU
 
+#ifdef CONFIG_KSCAN_FT5336_INTERRUPT
```
I see excessive use of "#ifdef & co" here. Please revise it.

dhoove replied:
I believe I cleaned it up about as much as I could.

Since there are three different cases, I thought having this block at the beginning of the code made sense, to determine early on:
- whether we need a timer or not
- on which IRQ edge(s) we must react

I did have a little extra clutter there, however. Thanks.

jfischer-no replied:
This is not resolved and I do not understand why you mark it as such. All these ifdef constructs here are not easy to read. I also have no understanding that one makes such an effort with macros just to save a few octets in RAM and flash. I think this driver should have only two options, interrupt mode and polling mode, or better just interrupt mode. And why do you need KSCAN_FT5336_PERIOD in the interrupt mode? Why can not interrupt mode not just always support up and down events?

dhoove replied:
Sorry if I clicked on 'resolve issue' when I shouldn't have. I was waiting for some feedback, and I though that maybe you were waiting for me to click on that in order to signal to you that I have addressed your concern.

I don't believe it's really possible to make this code simpler except by reducing the number of handled cases (which you mention). Currently, there are three cases:
- polling mode
- interrupt mode
- interrupt polling mode

Interrupt mode samples only at the highest possible sampling rate. For that reason, it is undesirable for low duty-cycle applications.

Polling mode has poor performance. But it should probably not be eliminated, because some hardware devices may not have a dedicated interrupt line.

Interrupt polling mode has configurable sampling rate, making it much more appropriate for low duty-cycle applications. I personally believe that interrupt mode is rendered obsolete by interrupt polling mode. But I didn't feel like it was up to me to make this decision and remove it. For this reason, I added the third case (interrupt polling mode), keeping the two others. This mode is by far the best solution for the case where only down and up events are necessary. It reacts immediately, and if the sampling period is set to 0, it almost always uses zero MCU time.

I don't understand why you say I made a massive effort with macros to save a few octets in RAM. I didn't do that at all. I simply added the third case in what I thought was a reasonably readable way, because I didn't feel like it was up to me to delete 'interrupt mode'. Of course, if you wish, I could remove the `KSCAN_FT5336_GPIO_INT_EDGE` and `KSCAN_FT5336_ENABLE_TIMER` defines which are the root cause of your apparent discontent. That would simply move the mess further down into the heart of the code, making easy-to-read code like the following much messier:
```
#ifdef KSCAN_FT5336_ENABLE_TIMER
	/** Timer (polling mode or interrupt polling mode). */
	struct k_timer timer;
#endif
```

Concerning your question about why KSCAN_FT5336_PERIOD exists in interrupt polling mode: it is used to select the sampling period. In 'interrupt mode', this is not configurable, and for hardware-related reasons, the device is always sampled at approximately 20ms.

In interrupt-polling mode, KSCAN_FT5336_PERIOD allows the developer to select an appropriate value to satisfy application requirements. If the developer selects zero, no sampling is done at all between down and up events, so the MCU use is almost always zero. Perhaps you could try out the code and compare the three modes while watching the SPI bus on a logic analyzer. The advantage of interrupt polling mode might become more clear to you.

What do you suggest I do? I personally believe that this code is an improvement over what was previously available. Would you like me to delete 'interrupt mode'? That would result in minor simplifications of the code, and I could then delete the additional #defines mentioned above that seem to disturb you. I don't really see the point. Despite the three cases, the file is reasonably readable. And there may be cases where the developer doesn't care about the MCU hogging that happens with interrupt mode, and wishes to sample at the highest possible sampling rate (I believe this could almost be achieved, however, by using interrupt polling mode with KSCAN_FT5336_PERIOD=1.

@MaureenHelm, @gmarull, do you believe that this driver handles too many cases, resulting in messy code? Should any of the three cases (polling, interrupt, interrupt polling) be removed?

gmarull reviewed (changes requested):
Thanks for these changes, having this mode will certainly help. I'd prefer to see the option available via DeviceTree, maybe a `int-polled` boolean option. All interrupt related Kconfig options could actually be deleted, I think. Current solution uses quite a lot of ifdefs, making code difficult to read.

dhoove commented on drivers/kscan/kscan_ft5336.c:
```diff
@@ -204,6 +245,13 @@ static int ft5336_init(const struct device *dev)
+	if (i2c_reg_write_byte(data->i2c, config->i2c_address, REG_G_MODE,
+			       REG_G_MODE_INTERRUPT_POLLING) < 0) {
+		LOG_ERR("Could not configure interrupt polling mode.");
+	}
```
Thanks.I had to declare ret, and I placed the declaration inside the #ifdef block to avoid potential compiler warnings.

dhoove replied:
Since the polling/interrupt behavior is driver-related, and not really device-related, are you sure it belongs in the device tree? Either way, I agree that there are a few too many compilation options. But I see cases where all three configurations more-or-less make sense...cheers.

dhoove reviewed (commented):
I believe I cleaned it up about as much as I could.

Since there are three different cases, I thought having this block at the beginning of the code made sense, to determine early on:

whether we need a timer or not
on which IRQ edge(s) we must react
I did have a little extra clutter there, however. Thanks.

dhoove reviewed (commented):
Hi. I made minor changes to your code recommendation to avoid compiler warnings and added declaration.

Since the polling/interrupt behavior is driver-related, and not really device-related, are you sure it belongs in the device tree? Either way, I agree that there are a few too many compilation options. But I see cases where all three configurations more-or-less make sense.

Repeating this since I'm not sure if you're waiting on changes from me. Thanks

gmarull reviewed (commented):
Changes LGTM, can you squash commits into a single one?

Commit:
drivers: kscan: add support for interrupt polling mode

Adds KSCAN_FT5336_INTERRUPT_POLLING_MODE to Kconfig: After interrupt
detects a down event, use timer-based polling until the up event. If
KSCAN_FT5336_PERIOD is zero, detect down and up events, but do not
poll in between.

Closes: #33994

Signed-off-by: David Hoover <github@hoover.li>

dhoove commented:
> Changes LGTM, can you squash commits into a single one?
Ok, done.

gmarull reviewed (approved):
can you rebase against `main` one more time (ci failure will likely go away)?

MaureenHelm reviewed (commented):
> Thanks for these changes, having this mode will certainly help. I'd prefer to see the option available via DeviceTree, maybe a int-polled boolean option. All interrupt related Kconfig options could actually be deleted, I think. Current solution uses quite a lot of ifdefs, making code difficult to read.

I agree the ifdefs make this hard to read. I think we could move the `REG_G_MODE` setting to device tree (i.e., `g-mode`), as it has a direct correlation to a device register. Many sensor drivers do this, and it allows different settings for each instance when there are multiple instances.

dhoove commented:
> Merge commit should be removed, just rebase against master branch. Suggested commit message: `drivers: kscan: ft5336: add support for interrupt polling mode`

Done. This has caused integration checks to fail in a way that seems unrelated to the code changes, and I'm receiving an email warning every hour or so...

dhoove commented:
> can you rebase against `main` one more time (ci failure will likely go away)?

Thanks, that worked.

dhoove commented:
@jfischer-no, Hi, I cleaned up the defines. Does this look okay?

dhoove commented:
@gmarull, @MaureenHelm, @jfischer-no:

Hi,

Is this blocked because of questions about the three configuration possibilities and a potential devicetree implementation?

Is there consensus about what needs to be done, or can we merge this and maybe add an issue suggesting further improvement?
